_posts/2022-05-27-pro-components-with-faasd.md

```diff
@@ -0,0 +1,367 @@
+```yaml
+faas-idler:
+    image: "ghcr.io/openfaasltd/faas-idler:0.5.0"
+    environment:
+      - "gateway_url=http://gateway:8080"
+      - "secret_mount_path=/run/secrets/"
+      - "prometheus_host=prometheus"
+      - "prometheus_port=9090"
+      # If a function is inactive for x minutes, it may be scaled to zero
+      - "inactivity_duration=3m" 
+      # The interval between each attempt to scale functions to zero
+      - "reconcile_interval=2m" 
+      # Write additional debug information
+      - "write_debug=false"
+    command:
+      - "/usr/bin/faas-idler"
+      - "-license-file=/run/secrets/openfaas-license"
+    volumes:
+      - type: bind
+        source: "./secrets/openfaas-license"
+        target: "/run/secrets/openfaas-license"
+      - type: bind
+        source: ./secrets/basic-auth-password
+        target: /run/secrets/basic-auth-password
+      - type: bind
+        source: ./secrets/basic-auth-user
+        target: /run/secrets/basic-auth-user
+    depends_on:
+      - gateway
+      - prometheus
+    cap_add:
+      - CAP_NET_RAW
+```
+
+One of the strenghts of faasd compared to OpenFaaS on Kubernetes is its sub second cold-start time. On Kubernetes scaling up form zero can take a couple of seconds. Faasd directly uses containerd which significantly reduces this cold start period.
+
+We can deploy the figlet function from the OpenFaaS store and add the required label.  
+```
+faas-cli store deploy figlet \
+  --label com.openfaas.scale.zero=true
+```
+
+After leaving the function idle for the configured time period it will be scaled to zero replicas.
+```
+Function                        Invocations     Replicas
+figlet                          1               0
+nodeinfo                        0               1 
+```
+
+When we call the function after it was scaled down the container will be restarted and once it is ready the function will be called. This all happens in under 90ms indluding the round trip latency.
+
+```
+time curl http://127.0.0.1:8080/function/figlet -d "Scale"
+ ____            _      
+/ ___|  ___ __ _| | ___ 
+\___ \ / __/ _` | |/ _ \
+ ___) | (_| (_| | |  __/
+|____/ \___\__,_|_|\___|
+                        
+
+real    0m0.085s
+```
+
+## Retrying requests
+It is possible to limit the amount of concurrent request that your function can process. This can be usfull when you are running memory intensive tasks and want to prevent overloading the function. Limiting requests can be done by setting `max_inflight` on the function. When a function cannot accept any more connections it will return a http status code 429. You could handle these types of errors by implementing your own retry machnism which but that would probabily be a tedious process. OpenFaaS Pro comes with an upgraded queue-worker that will retry messages a number of times using an exponential back-off arlgorithm. The pro queue-worker will retry messages a set number of times before dropping them. 
```

Comment:
```suggestion
It is possible to limit the number of concurrent request that your function can process. This can be useful when you are running memory intensive tasks and want to prevent overloading the function. Limiting requests can be done by setting `max_inflight` on the function. When a function cannot accept any more connections it will return a http status code 429. You could handle these types of errors by implementing your own retry mechanism which would probably be a time consuming and tedious process. OpenFaaS Pro offers an upgraded queue-worker that will retry messages a number of times using an exponential back-off algorithm. The pro queue-worker will retry messages a set number of times before dropping them. 
```